Dynamic token list support for publishing

LucaMilanese90 opened the pull request:
Fixes #1335

Changes proposed in this PR:

- Add support for approved baseTokens

mihaisc commented:
i'll do a quick pr to solve this https://github.com/oceanprotocol/ocean-subgraph/issues/355 so you can display token name, not address

LucaMilanese90 commented:
That would be great actually, can you also make the decimals available? This address has 6 digits which creates a lot of problems during publish

mihaisc commented:
yes, it will be a 'Token' object

mihaisc commented:
fixed and deployed 
![image](https://user-images.githubusercontent.com/26000280/173401780-6ec96402-23c6-42ee-ad1c-7c502ba55a7b.png)

LucaMilanese90 commented:
@mihaisc I updated the query to display the symbols instead of the addresses and I'm updating the labels accordingly during the publish flow. I'm able to publish with fix price (pool still gives errors) but trying to buy does not work. Is there something else I have to change other than baseTokenAddress and decimals?

mihaisc commented:
hmm, it should just work, i published some time ago with tusd and it just worked. I'll do some tests.

LucaMilanese90 commented:
Have you published with pool from the preview link of from some local branch? I keep getting the same error
<img width="572" alt="Screenshot 2022-06-14 at 11 53 43" src="https://user-images.githubusercontent.com/20192135/173549633-3cd02446-1193-40cf-b690-4f5fa799e4df.png">

mihaisc commented:
directly from preview, do you have tusdt in your wallet? not sure

LucaMilanese90 commented:
Yep, I have 10, very strange🤔

<img width="358" alt="Screenshot 2022-06-14 at 13 52 35" src="https://user-images.githubusercontent.com/20192135/173570893-406c354a-81a4-4560-87f1-b46cade1616c.png">

mihaisc commented:
but how to you want to create a pool with only 10?

LucaMilanese90 commented:
Yes, I  realized right after I wrote the comment, I thought I had 100 🤦‍♂️

LucaMilanese90 commented:
That's because the balance in the Web3 context provider is only checking for eth and ocean, I'm changing the Web3 context provider to also include the other tokens

LucaMilanese90 commented:
When publishing with pool price the weight of the baseToken is '5000000000000' instead of '5' when using TUSDT (the 12 decimals of difference between TUSDT and OCEAN), but the decimals seem to be set correctly in the publish phase, any ideas?

mihaisc commented:
I don't think i understand.

mihaisc commented:
The base token weight will be fixed here https://github.com/oceanprotocol/ocean-subgraph/pull/468 . Is there anything else left to do here? Seems good

LucaMilanese90 commented:
@mihaisc I just pushed the last changes to show the balance for the correct token in pool add/remove and trade tabs. Add/remove and trade seems to work correctly, only buy (for fixed and dynamic) is not currently working correctly. 
How do we deal with the text on the publish flow? Do we want to replace OCEAN with a more neutral text or is it ok to keep it like this?
<img width="772" alt="Screenshot 2022-06-23 at 17 43 23" src="https://user-images.githubusercontent.com/20192135/175340008-29a40721-3b91-4afe-9dfd-dc3d58f3c97d.png">

And what about the Ocean logo next to the basetoken? Is that ok or should we use a different image for the alternatives?
<img width="361" alt="Screenshot 2022-06-23 at 17 43 36" src="https://user-images.githubusercontent.com/20192135/175340361-8f3a8ce7-3474-46ca-8ec6-477c03680c41.png">

mihaisc commented:
>  buy (for fixed and dynamic) is not currently working correctly.

I'll do some debugging as well, we also had some issues with the provider at the time you were testing , maybe that was the problem.

We should replace `OCEAN` with a more neutral text like `selected base token` 

We are using H2O so you can leave the same image. H2O doesn't have an actual image

mihaisc commented:
Please update ocean.js to 1.1.3 , it should fix ordering

LucaMilanese90 commented:
Still not working, now I get an error as soon as I click "buy" both on fixed and dynamic price assets. 
<img width="508" alt="Screenshot 2022-06-25 at 10 26 47" src="https://user-images.githubusercontent.com/20192135/175765412-283d7bdd-d3a3-4b85-b162-1cee0e7df31c.png">
 
It looks like the price is not calculated correctly, for example on this asset did:op:286af94120a77fb3cfae56cbe37789a44c40d86d9ec90b2c711ffc7dd16bdfe9 I see the tokenAmount is "0.000000000001012125", which is not correct.

mihaisc commented:
Ok, i'll take a look.

mihaisc commented:
All good now. We were setting the wrong decimals for the datatoken in publish. So you need to publish new assets to test.

LucaMilanese90 commented:
@mihaisc amazing! Tried with fixed and pool and the buying process is working correctly. Same for adding/removing from the pool. The only thing that I can see it's still failing is trading datatokens for basetokens, which is strange because the opposite seems to be working correctly (not sure if it's related to the fact the basetokenWeight is still "50000000000000") on an [asset](https://market-x31mi2462-oceanprotocol.vercel.app/asset/did:op:ead984089c868121b22d4a5e2c9ba53664b2f0eff5cfcd2bea3c262e77e456e3) published from the new deployed preview 🤔

mihaisc commented:
That's probably normal since i didn't deploy the subgraph 😁. I'll do all the updates tomorrow

LucaMilanese90 commented:
Ok, then I guess we are ready for review?

mihaisc commented:
All good now.

EnzoVezzaro reviewed (approved):
LGTM

EnzoVezzaro reviewed (dismissed):
Wait. Not really.

<img width="452" alt="Screenshot 2022-06-29 at 11 16 57" src="https://user-images.githubusercontent.com/13741335/176476073-2b4359cb-94b7-4591-8f75-aac228711473.png">

The message at the bottom should have TUSDT instead of OCEAN in [this case](https://market-1zxmteee6-oceanprotocol.vercel.app/asset/did:op:ead984089c868121b22d4a5e2c9ba53664b2f0eff5cfcd2bea3c262e77e456e3).

LucaMilanese90 commented:
> Wait. Not really.
> 
> <img alt="Screenshot 2022-06-29 at 11 16 57" width="452" src="https://user-images.githubusercontent.com/13741335/176476073-2b4359cb-94b7-4591-8f75-aac228711473.png">
> 
> The message at the bottom should have TUSDT instead of OCEAN in [this case](https://market-1zxmteee6-oceanprotocol.vercel.app/asset/did:op:ead984089c868121b22d4a5e2c9ba53664b2f0eff5cfcd2bea3c262e77e456e3).

Should be fixed now

kremalicious commented:
Why is BaseToken this huge new single `select` form field? Where we then only have 2 options?

This should be just a `select` on the price input field prefix, where instead of OCEAN I can select the other base tokens. Selecting this then switches base token across whole pricing form

Select is just empty without a wallet, and tokens are wrong too, showing "DT" and "Data Token" on OCEAN field

<img width="1004" alt="Screen Shot 2022-06-30 at 11 12 55" src="https://user-images.githubusercontent.com/90316/176652609-b21730d2-e2c5-4677-9000-fc849905f00a.png">

After connecting wallet on this screen as I am prompted there, I get this, with something selected in that `select` at top, which then is not reflected in the form:

<img width="1030" alt="Screen Shot 2022-06-30 at 11 15 09" src="https://user-images.githubusercontent.com/90316/176653130-aa6cf07e-7308-40d3-a36d-bdfcb6c8cc26.png">

I first have too select OCEAN in that `select` again, then switch back to H2O. Only then changes are reflected

mihaisc commented:
There is no h20 logo

kremalicious commented:
How to test H2O publishing? Only appears for me on main networks, but not on any test network. If there is only 1 approved token, the UI becomes even more redundant, where in case of making the input prefix into a `select` we can simply hide the caret in this case to remove confusion. In this current UI iteration this is not possible

kremalicious commented:
> There is no h20 logo

well, but we can agree that the Ocean logo is not the H2O logo? And Coingecko shows something for me: https://www.coingecko.com/en/coins/h2o

kremalicious reviewed (changes requested):
remove redundant base token `select` field and deal with some visual and interaction weirdnesses described in former comments

LucaMilanese90 commented:
Went for the big select because the only other select implementation on the publish flow is the "timeout" on step 2, so I thought you wanted to keep the same style for this one as well. But no problems, I'll move it to the price label.
As for the different logo based on the token, how can I query for the appropriate images?

kremalicious commented:
> As for the different logo based on the token, how can I query for the appropriate images?

That would need to be figured out. If the goal here is to launch as quickly as possible then would suggest dealing with this manually for now and copy over some SVG assets for the first approved tokens only.

For full automation, there are a bunch of token lists out there where projects self-maintained them and apps then use those token lists to get metadata about tokens. Like this one https://github.com/trustwallet/assets

This would then also require to deal with multiple asset types then, as images could be bitmap or vector assets so that all becomes a bit more tricky to implement. Would say 3-5 approved tokens is a good baseline after which we should implement more automatic way. Until then copying in assets should be enough.

kremalicious commented:
code archaeology from coin selection in v3:

https://github.com/oceanprotocol/market/blob/9f2cbaafc1fbae74303ef8b478677a9a46d0d385/src/components/organisms/AssetActions/Pool/Add/FormAdd.tsx#L137-L156

https://github.com/oceanprotocol/market/blob/9f2cbaafc1fbae74303ef8b478677a9a46d0d385/src/components/organisms/AssetActions/Pool/CoinSelect.tsx

https://github.com/oceanprotocol/market/blob/9f2cbaafc1fbae74303ef8b478677a9a46d0d385/src/components/organisms/AssetActions/Pool/CoinSelect.module.css

mihaisc commented:
My bad on the image i found another H2O and when i asked i was  told there is none and dropped it.
Agree with everything. If the image implementation is too complicated just stick with a hardcoded array/list and add the required images in whatever format we desire. We are the ones who add approved tokens, and we don't plan more for now. So doing a very complicated implementation would not really make sense

kremalicious commented:
some more minor UI thingies:

found another mention of OCEAN hardcoded, also not sure if it's of any concern because of test token, but the conversion into user currency would be wrong for a supposed stable coin. Although stable coins loosing their peg are hip right now

<img width="780" alt="Screen Shot 2022-06-30 at 21 05 07" src="https://user-images.githubusercontent.com/90316/176768135-cf84b869-d695-4824-a84a-44f46cb01942.png">

The error validation message overlapping the input bug is back:

<img width="549" alt="Screen Shot 2022-06-30 at 21 07 20" src="https://user-images.githubusercontent.com/90316/176768417-6d5a538d-8adb-4da0-9d12-dca4fa3e63c3.png">

This is only in this PR, not present on live where it behaves as intended:
<img width="449" alt="Screen Shot 2022-06-30 at 21 09 03" src="https://user-images.githubusercontent.com/90316/176768660-e7f93eed-d52a-44af-8181-7cbada4f93bd.png">

That balance check message in last screenshot appears and reappears sporadically for me when I have none of that token in my wallet. But we have already WIP PR going on for it with #1435 so might be rather solved there.

EnzoVezzaro commented:
If I'm not connected I just see an empty input. It's kind of weird to see. Maybe we add a placeholder instead.

![Screenshot 2022-07-04 at 07 59 09](https://user-images.githubusercontent.com/13741335/177150603-a821a188-c34c-4a70-8961-ba273466d350.png)

kremalicious commented on src/@context/Web3.tsx:
```diff
@@ -155,16 +157,33 @@ function Web3Provider({ children }: { children: ReactNode }): ReactElement {
     try {
+      const { oceanTokenAddress } = getOceanConfig(networkId)
+      const approvedTokensBalance: { [key: string]: string } = {}
+      const approvedTokens = (await getOpcsApprovedTokens(chainId)).filter(
```
I see we call `getOpcsApprovedTokens` multiple times in various places, again on the pricing page. Would it be more effective if we fetch and expose approved tokens from `MarketMetadata`? Cause this Web3 context provider here can then just grab it from there, as `MarketMetadata` is our highest context provider. Cause in end we should only fetch the approved tokens once, when user initially enters app

LucaMilanese90 replied:
Yes, that is a good idea actually! I'll modify it as suggested

LucaMilanese90 replied:
In that provider I'm not going to be able to check for the selected chainId though, right? Should I expose the list for all the supported chainIds then?

kremalicious replied:
right, didn't think of that and if I count the requests current implementation could be actually most effective already. Think we're good here, sorry!

Cause if we always fetch for all chains on app start we force at least 5 or 9 requests on all users although so far they are only relevant in the publish form. But if e.g. user is connected to Polygon and publishes on Polygon we end up with 2 requests (1 in Web3 context, 1 in pricing form). We could reduce this to only 1 by storing the fetched tokens in some context but too much refactor for this PR

kremalicious commented:
awesome, thanks for all the latest changes, much better UI and interactions now. Could only find this on the first time user enters this screen, where it says "Data Token" where it should say "Ocean Token":

<img width="903" alt="Screen Shot 2022-07-05 at 12 57 18" src="https://user-images.githubusercontent.com/90316/177322114-bcbf7517-1ef0-4040-925b-541fcbd9ed31.png">

Works correctly when switching to H2O, then back to OCEAN.

Then was already trying to ditch heading font for token names but then realized there is actually a real space in code and not just some bad kerning:

<img width="448" alt="Screen Shot 2022-07-05 at 12 51 13" src="https://user-images.githubusercontent.com/90316/177321132-bc54f847-f5b2-4513-a5eb-d770968d5f72.png">

As for the overlapping error state, this is because when H2O is selected the conversion div under the input field is gone, moving the absolute positioned error up. So if we can get the conversion to show up on H2O, we don't need to do anything regarding error placement, if we want to leave it out for now, guess we have to do some CSS tweaks for that error alone.

As for conversion in general I can look into teaching our price context provider more tokens but would do that in another PR as most likely will change lots of files. We can start with hardcoded list as we would have to map approved tokens to Coingecko token ids but as long as token is listed there this should work out

LucaMilanese90 commented:
> awesome, thanks for all the latest changes, much better UI and interactions now. Could only find this on the first time user enters this screen, where it says "Data Token" where it should say "Ocean Token":

Yep, I'm still working on it, I'll fix this😉

LucaMilanese90 commented on src/components/Publish/Pricing/index.tsx:
```diff
@@ -11,75 +11,123 @@ import Free from './Free'
+    pricing
+
+  useEffect(() => {
+    if (!chainId) return
```
Not 100% sure why but with `(chainId || 1)` everything breaks and I always get back the results from chainId = 1 🤔

LucaMilanese90 commented:
> Then was already trying to ditch heading font for token names but then realized there is actually a real space in code and not just some bad kerning:

Can I simply use the string in `token.name` then? The problem is that while for ETH is fine and the name is 'Ocean Token' 
<img width="460" alt="Screenshot 2022-07-05 at 14 54 01" src="https://user-images.githubusercontent.com/20192135/177332334-65054d78-7091-41a1-82f3-1123c4914f8a.png">

for the testnets it's not, on Rynkeby for example it is 'oceanToken'
<img width="454" alt="Screenshot 2022-07-05 at 14 54 45" src="https://user-images.githubusercontent.com/20192135/177332483-d67cded3-b63b-4ebf-8841-92cedd861b68.png">

I tried to convert the camelCase string with regex but there are too many edge cases to handle.

Also, would it be safe for me to assume that the default token is always the one in position `0` from the approved tokens array? Or are they ordered alphabetically?

kremalicious commented:
> Can I simply use the string in token.name then?

yes! No need to complicate this, wasn't sure where exactly `h2 o` was coming from so that's the only concern regarding spacing. Your screenshot is technically correct, as on Rinkeby the token name in the deployed contract is indeed `OceanToken` as we never deployed the token update there we had on mainnet, where we also changed the name to `Ocean Token`. In any case, think it's fine just passing through `token.name`

Over in #1573 made the prices context more token-agnostic so we can deal with that token conversion topic later over there.

LucaMilanese90 commented:
@kremalicious update:

> As for the overlapping error state, this is because when H2O is selected the conversion div under the input field is gone, moving the absolute positioned error up. So if we can get the conversion to show up on H2O, we don't need to do anything regarding error placement, if we want to leave it out for now, guess we have to do some CSS tweaks for that error alone.

The error positioning should be fixed now even if the conversion is not shown.

> in this case, we should have a `chainId` fallback so that for demonstration reasons we also fetch when no wallet is connected. In this case the whole publish form defaults to `1` so if we remove this here and further down do `await getOpcsApprovedTokens(chainId || 1)`, then we have this functionality for the no wallet case too

This is now implemented, I've also added a check to verify if the chainId is supported, if not it's reverting back to the default 1

As for the token logos, should the color scheme remain the same (black for basetoken and pink for datatoken)?
<img width="750" alt="Screenshot 2022-07-06 at 10 39 34" src="https://user-images.githubusercontent.com/20192135/177509189-6ce5c770-ba03-43c8-bffc-f5360163e892.png">

These logos are updated as well (don't mind the wrong token name, I couldn't find a published asset with H20 so I just switched the logos for the screenshot)

<img width="335" alt="Screenshot 2022-07-06 at 10 46 14" src="https://user-images.githubusercontent.com/20192135/177509875-af276397-db48-4acb-aaf6-f4ece6c0dbfd.png">


Is the logo in the wallet supposed to stay ocean or should that change too? 
<img width="153" alt="Screenshot 2022-07-06 at 10 50 58" src="https://user-images.githubusercontent.com/20192135/177510692-591017ff-4012-43bf-ada1-d5817f15ecb4.png">

mihaisc commented:
> Is the logo in the wallet supposed to stay ocean or should that change too?

Leave it as it is for now, we need to rethink that a bit.

kremalicious commented on src/@context/Web3.tsx:
```diff
@@ -155,16 +157,33 @@ function Web3Provider({ children }: { children: ReactNode }): ReactElement {
       const balance = {
         eth: web3.utils.fromWei(await web3.eth.getBalance(accountId, 'latest')),
-        ocean: await getOceanBalance(accountId, networkId, web3)
+        ocean: await getOceanBalance(accountId, networkId, web3),
```
bonus points for not making duplication in codebase worse: `getOceanBalance()` from `@utils/ocean` is redundant and unnecessarily specific, so can be replaced completely with the newly created `getTokenBalance()`.

LucaMilanese90 replied:
I refactored the user balance fetch and took the occasion to move the approvedBaseTokens list into the web3 provider to reduce the number of calls of the `getOpcsApprovedTokens()` function.

kremalicious commented on src/components/Header/Wallet/Details.tsx:
```diff
@@ -67,7 +67,7 @@ export default function Details(): ReactElement {
                 address={oceanTokenMetadata?.address}
                 symbol={oceanTokenMetadata?.symbol}
-                logo="https://raw.githubusercontent.com/oceanprotocol/art/main/logo/token.png"
+                logo="ocean"
```
up here, the logo is supposed to be for the datatoken, as this action adds the datatoken to the wallet. Just like with pricing screen, this raises question what exact logo datatokens should have, as this was never consciously decided. I would see 3 options:

1. Giving datatokens the ocean logo, but in violet, was a quick solution when building v3, but it also reflects that they are still OCEAN-branded tokens in a way. From that perspective we could go with violet OCEAN logo for all datatokens, regardless of base token.

2. Current implementation where base token logo is made violet works also well if we are fine with them being branded differently. But mechanics-wise the datatoken as access token is not bound or derived from the base tokens so argument could be made this could be confusing to users.

4. We could come up with a new logo, for all datatokens. Of course would take longest as design work has to be done first.

mihaisc replied:
3 would be the best , but i would go with 1. for now.

LucaMilanese90 replied:
So, are we going with solution 1 for the moment? Can I go ahead and change it?

mihaisc replied:
go for it

LucaMilanese90 replied:
Done

mihaisc commented:
On polygon where do we get that name from ? `H20data` . And i assume the icon doesn't change because of the name

![image](https://user-images.githubusercontent.com/26000280/177975989-15975b64-f379-4e0b-85cf-712c8391ce06.png)

LucaMilanese90 commented:
> On polygon we display `H20data` . I assume the icon doesn't change because of the name ![image](https://user-images.githubusercontent.com/26000280/177976351-825b0592-b10d-4e7f-8143-5f0d821aec39.png) I think we should use the symbol. Looks on all networks

Looks lithe the problem is that on Ethereum the symbol is H2O (with the letter "o" at the end), while on polygon it's spelled with the number 0.

mihaisc commented:
Ok , i think everything is covered. @kremalicious see any major things that are missing?

EnzoVezzaro commented:
Just [published](https://market-5sqxkn7eu-oceanprotocol.vercel.app/asset/did:op:950293cefa3fe53cd15040564ceb0a6e8fcd40a84f77172ce3a042c41010f8e0) on polygon, I was able to publish with h2o, but the details page breaks because some data is missing

<img src='[figure/jupyterlab.png](https://user-images.githubusercontent.com/13741335/179526606-a089a395-641f-463a-9aec-3b8166104499.png)' width='250'>

On the response I see missing some info in baseToken

![Screenshot 2022-07-18 at 09 55 08](https://user-images.githubusercontent.com/13741335/179527374-4eb086de-702d-4c5c-8881-12f4eaa64223.png)

- on edit I see `Ocean` instead of `h20`

![Screenshot 2022-07-18 at 09 54 06](https://user-images.githubusercontent.com/13741335/179527883-fb6ce639-2691-4bd0-9df2-66db71913713.png)

- on the publish form found an issue on the checkbox of the pricing tab

`Fixed` is fine
![Screenshot 2022-07-18 at 09 48 34](https://user-images.githubusercontent.com/13741335/179527615-482ac18f-3ef5-4c53-82e7-bd9c462f6f6c.png)

`Free` is missing filled bullet
![Screenshot 2022-07-18 at 09 48 39](https://user-images.githubusercontent.com/13741335/179527617-db985dcb-4a8b-4ed8-a2e7-e29559123cfc.png)

mihaisc commented:
The problem is that on mumbai we have only 1 token [query](https://v4.subgraph.mumbai.oceanprotocol.com/subgraphs/name/oceanprotocol/ocean-subgraph/graphql) so don't understand why there are two in the publish form, will check it out

LucaMilanese90 commented:
I only see 1 token when connected to Mumbai (don't see the dropdown at all). I think the problem is if you compile the form without being connected with your wallet and then connect after the price step: since it was decided to default to `chainId` 1 to fetch the approved baseTokens when no wallet is connected, you'll always be able to see H20 and Ocean when disconnected. We could reset the baseToken value when the `chainId` changes at any point during the publish flow or check during the creation of the nft if the baseToken address is coherent with the selected network. Which option would you prefer?

mihaisc commented:
>  We could reset the baseToken value when the chainId changes at any point during the publish flow or check during the creation of the nft if the baseToken address is coherent with the selected network.

this sounds good

LucaMilanese90 commented:
Which one? Or do you mean both?

mihaisc commented:
omg... sorry 

> We could reset the baseToken value when the chainId changes at any point during the publish flow

AnaLoznianu commented:
Please keep in mind that the AMM pool removal will be merged in the main branch on the 1st of August and we will merge this PR soon after.

kremalicious commented:
Super big bug, on Polygon the symbol uses a `0` instead of a big `O`

<img width="163" alt="Screen Shot 2022-07-27 at 16 46 05" src="https://user-images.githubusercontent.com/90316/181291286-3defaff9-b0f5-4734-9df7-923502cbc841.png">

Another weird but small interaction bug: select `H2O`, then go to next screen. Then go back, and the form data will be reset to the initial default base token.

Apart from that, all seems to be working well, great job!

Remaining conversion issue then can be integrated and finalized in #1573

LucaMilanese90 commented:
> > On polygon we display `H20data` . I assume the icon doesn't change because of the name ![image](https://user-images.githubusercontent.com/26000280/177976351-825b0592-b10d-4e7f-8143-5f0d821aec39.png) I think we should use the symbol. Looks on all networks
> 
> Looks lithe the problem is that on Ethereum the symbol is H2O (with the letter "o" at the end), while on polygon it's spelled with the number 0. I added these options to the TokenLogo switch case.

@kremalicious yes, I mentioned it in an old comment: the H2O written in two different ways is the result of the approved token query, we can manually override it in the frontend but it would be better to have the correct data from the query itself. 

On the default token, up until now, it was decided to use the first token in the list as default for each network. We can change that and use a `.find()` or something else to catch the ocean token from the list.

You are right on the weird reset of the default token when going back and forth between steps, that should be an easy fix.

mihaisc commented:
> Super big bug, on Polygon the symbol uses a 0 instead of a big O

Not a bug, that is how it's on the chain https://polygonscan.com/address/0xc5248aa0629c0b2d6a02834a5f172937ac83cbd3

LucaMilanese90 commented:
Update:
- the default baseToken is set to look for the ocean option
- the token reset bug when switching tabs back and forth should now be fixed (it will still reset if the chainId changes during the publish process)

kremalicious commented on src/components/Publish/Pricing/Price.tsx:
```diff
@@ -59,12 +58,17 @@ export default function Price({
-              <Conversion price={field.value} className={styles.conversion} />
-            </h4>
           </div>
+          {firstPrice && (
```
there is no `firstPrice` anymore so this can be removed again, same for respective CSS rules.

mihaisc replied:
cool ,didn't know what was that for

jamiehewitt15 commented:
In terms of the visual aspect, the grey background makes the "OCEAN" look like something that can't be changed (like the fee fields beneath it)

![H2O - 2022-08-02_16-08](https://user-images.githubusercontent.com/20739535/182382664-a436e031-cdc0-443b-9fd9-be4d2603fe36.png)

By contrast, in other drop down boxes that we allow people to change the background is white. 
![Timeout - 2022-08-02_16-08](https://user-images.githubusercontent.com/20739535/182382567-10f0407d-150e-424c-9de0-e9a2f95970c5.png)

I would suggest changing the background colour to white to bring more attention to it.

jamiehewitt15 commented:
Another potential improvement would be to put a tool tip next to pricing so that we can explain briefly what H2O is. As it stands, there is no explanation about H2O or any links to where users can get more info. At the moment we don't have anything in the docs about it either (I'll create an issue for that). 

The functionally works well though, I haven't been able to break it or find any bugs.

mihaisc commented:
The docs are enough, no more tooltips required.

bogdanfazakas commented:
Should small issues on c2d be fixed before merging or we can fix them later on since C2D  is still in a testing phase ?
In this case there is a problem at display when we mix dataset and algorithm with diferent tokens (dataset is TUSDT but algo is OCEAN in the case below)

<img width="441" alt="Screenshot 2022-08-03 at 09 50 20" src="https://user-images.githubusercontent.com/13447142/182543392-c81fe269-3804-4fc5-957c-b30fde4214db.png">

mihaisc commented:
we can fix them after we merge in a separate pr

kremalicious commented:
> the grey background makes the "OCEAN" look like something that can't be changed

code to make its background light is actually there but it won't kick in as we are using a `background` later in the code to create the caret without any images. On top of that this is our input prefix component so we should not change them all to white, as we only want more affordance in it if it is an actual `select`. Which also won't be all the time in that price field, like on network where only OCEAN is supported.

So as compromise so we don't have to refactor all input stuff just for this:

Only 1 option:

<img width="468" alt="Screen Shot 2022-08-03 at 12 43 57" src="https://user-images.githubusercontent.com/90316/182599791-7f443a9a-1811-4bb4-b99c-ea6a9cbb5e13.png">

With multiple options:

<img width="440" alt="Screen Shot 2022-08-03 at 12 45 09" src="https://user-images.githubusercontent.com/90316/182599998-300404e1-f79e-4a71-b29c-afe5e80b8a79.png">

kremalicious commented:
In more surprising problems, I am the one with this last remaining conversation but I can't get to it as clicking the View link takes me to long loading changeset view but this conversation to resolve is nowhere to be found.

<img width="879" alt="Screen Shot 2022-08-03 at 12 46 59" src="https://user-images.githubusercontent.com/90316/182600330-bbfc4ff6-cbd3-4b56-aec7-69d7197b6bbe.png">


So we might have to quickly change merge permissions and back to get this merged, also because @EnzoVezzaro has an ongoing change requests which also prevents merge unless resolved.
